ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/subject/confirmations/BearerSubjectConfirmationVerification.kt

```diff
@@ -0,0 +1,126 @@
+import org.codice.compliance.SAMLProfiles_4_1_4_2_k
+import org.codice.compliance.attributeNode
+import org.codice.compliance.attributeText
+import org.codice.compliance.children
+import org.codice.compliance.utils.ASSERTION
+import org.codice.compliance.utils.AUDIENCE
+import org.codice.compliance.utils.AUTHN_STATEMENT
+import org.codice.compliance.utils.BEARER
+import org.codice.compliance.utils.METHOD
+import org.codice.compliance.utils.SUBJECT
+import org.codice.compliance.utils.SUBJECT_CONFIRMATION
+import org.codice.compliance.utils.SUBJECT_CONFIRMATION_DATA
+import org.codice.compliance.utils.TestCommon
+import org.codice.compliance.utils.TestCommon.Companion.REQUEST_ID
+import org.codice.compliance.utils.TestCommon.Companion.currentSPIssuer
+import org.codice.compliance.utils.TestCommon.Companion.idpMetadata
+import org.codice.security.saml.SamlProtocol.Binding.HTTP_POST
+import org.w3c.dom.Node
+
+class BearerSubjectConfirmationVerification(private val samlResponseDom: Node) {
+    fun verify() {
+        verifyBearerSubjectConfirmations()
+        verifyBearerAssertions()
+        verifyAudienceRestriction()
+    }
+
+    @Suppress("ComplexCondition")
+    private fun verifyBearerSubjectConfirmations() {
+        val bearerSubjectConfirmationsList = mutableListOf<Node>()
+        samlResponseDom.children(ASSERTION).forEach {
+            val bearerSubjectConfirmations = it.children(SUBJECT)
+                .filter { it.children(SUBJECT_CONFIRMATION).isNotEmpty() }
+                .flatMap { it.children(SUBJECT_CONFIRMATION) }
+                .filter { it.attributeText(METHOD) == BEARER }
+
+            if (bearerSubjectConfirmations.isEmpty())
+                throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_g,
+                    message = "An assertion with no bearer SubjectConfirmation elements was found.",
+                    node = it)
+
+            bearerSubjectConfirmationsList.addAll(bearerSubjectConfirmations)
+        }
+
+        if (bearerSubjectConfirmationsList
+                .filter { it.children(SUBJECT_CONFIRMATION_DATA).isNotEmpty() }
+                .flatMap { it.children(SUBJECT_CONFIRMATION_DATA) }
+                .none {
+                    it.attributeText("Recipient") ==
+                        TestCommon.getServiceUrl(HTTP_POST, samlResponseDom) &&
+                        it.attributeNode("NotOnOrAfter") != null &&
+                        it.attributeNode("NotBefore") == null &&
+                        it.attributeText("InResponseTo") == REQUEST_ID
+                }) {
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_h,
+                message = "There were no bearer SubjectConfirmation elements that matched " +
+                    "the criteria below.",
+                node = samlResponseDom)
+        }
+    }
+
+    private fun verifyBearerAssertions() {
+        val bearerAssertions = samlResponseDom.children(ASSERTION)
```

Comment:
All assertions in the SSO profile are Bearer Assertions because:
```
Any assertion issued for consumption using this profile MUST contain a <Subject>
element with at least one <SubjectConfirmation> element containing a Method of
urn:oasis:names:tc:SAML:2.0:cm:bearer.
```